### PR TITLE
WAZO-889-integrate-microsoft-plugin

### PR DESCRIPTION
**Pull request by pc-m**

### CODE REVIEW

File: wazo_dird/plugins/office365_backend/plugin.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+        for contact in updated_contacts:
+            if self._first_match_predicate(lowered_term, contact):
+                return self._SourceResult(contact)
+
+    def _first_match_predicate(self, term, contact):
+        for column in self._first_matched_columns:
+            column_value = contact.get(column) or ''
+
+            if not isinstance(column_value, list):
```
Comment (sileht) [inline wazo_dird/plugins/office365_backend/plugin.py]: Starting with the True expression first is more intuitive, e.g:

`if isinstance(column_value, list):`

File: wazo_dird/plugins/office365_backend/schemas.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+class SourceSchema(BaseSourceSchema):
+
+    auth = fields.Dict(
+        keys=fields.String(validate=Length(min=1, max=64)),
+        values=fields.String(validate=Length(min=1, max=64))
+    )
+    endpoint = fields.String(
+        missing='https://graph.microsoft.com/v1.0/me/contacts',
+        validate=Length(min=1, max=255),
```
Comment (sileht) [inline wazo_dird/plugins/office365_backend/schemas.py]: Does Microsoft provides this service on premise or we have something only for testing in our API ?

File: wazo_dird/plugins/office365_backend/services.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            "search": term
+        }
+        try:
+            response = requests.get(url, headers=headers, params=query_params)
+            if response.status_code == 200:
+                logger.debug('Sucessfully fetched contacts from microsoft.')
+                return response.json().get('value', [])
+            else:
+                return []
+        except requests.exceptions.RequestException as e:
```
Comment (sileht) [inline wazo_dird/plugins/office365_backend/services.py]: `requests.RequestException` is enough if you want to save some char.

File: wazo_dird/plugins/office365_backend/plugin.py (hunk 1 above)
Comment (pc-m) [reply wazo_dird/plugins/office365_backend/plugin.py]: I agree

File: wazo_dird/plugins/office365_backend/schemas.py (hunk 2 above)
Comment (pc-m) [reply wazo_dird/plugins/office365_backend/schemas.py]: I don't know if Microsoft offers this service on premise. I think the parameter was there when the proof of concept was written, then it got into the tests and made it to production.

I would not mind replacing the argument with a config option to make the API simpler in the future.

**Comment by pc-m** [comment]
> Oh I just see that debian new dependencies are missing.

I add python3-marshmallow

**Comment by pc-m** [comment]
> requirements.txt need to be updated too.

I cannot find which ones are missing for dird, I will check wazo-auth

**Pull request merged**